[naga/spv-in] Remove unnecessary "gl_PerVertex" name check so unused builtins will always be skipped

Imberflur opened the pull request:
**Connections**
Fixes https://github.com/gfx-rs/wgpu/issues/4915

**Description**
Unused gl_PerVertex builtins need to be skipped since they my otherwise require capabilities that are not enabled. However, this was not being done when the input spirv didn't explicitly name this struct.

This just removes the check against the "gl_PerVertex" name as suggested in https://github.com/gfx-rs/wgpu/issues/4915#issuecomment-1926783047. Assuming valid SPIRV, this is technically equivalent to https://github.com/gfx-rs/wgpu/issues/4915#issuecomment-1873701825 because a struct with builtin members [must be decorated as a `Block`](https://github.com/KhronosGroup/SPIRV-Tools/commit/df2aad68b98279412494a6d449bd71b6756e699b) 

Would it be worth adding a test for this? If so, would that go under `naga/tests/in/spv`?

Also, I'm not very familiar with the overall structure of SPIRV nor with the spv frontend for naga. Can parse_function run more than once for a particular input? If so, is it possible that later functions will be using the builtin but we remove the builtin label after not seeing any uses when processing the first function?

**Testing**
I produced a .spv file using the method described in https://github.com/gfx-rs/wgpu/issues/4915 and confirmed that this change removes the error there. I also tested an application using `shaderc` to generate SPIRV for `wgpu` that was seeing this error with certain configurations of `shaderc` and confirmed that the issue was resolved (and that everything seemed to work).

<!-- 
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're no bothering us!
-->

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] ~~`--target wasm32-unknown-unknown`~~
  - [ ] ~~`--target wasm32-unknown-emscripten`~~
- [x] Run `cargo xtask test` to run tests.
  <details><summary>Some failures but they also fail on `trunk`</summary> 
  
  ```
  Summary [   6.158s] 411 tests run: 406 passed (5 flaky), 5 failed, 0 skipped
       FLAKY 2/3 [   0.251s] wgpu-examples [Executed] [Gl/AMD Radeon RX 6500 XT (radeonsi, navi24, LLVM 16.0.6, DRM 3.57, 6.7.4-arch1-1)/1] wgpu_examples::hello_compute::tests::compute_1
       FLAKY 2/3 [   0.374s] wgpu-examples [Executed] [Gl/AMD Radeon RX 6500 XT (radeonsi, navi24, LLVM 16.0.6, DRM 3.57, 6.7.4-arch1-1)/1] wgpu_examples::hello_compute::tests::compute_overflow
       FLAKY 2/3 [   0.271s] wgpu-test::wgpu-test [Executed] [Gl/AMD Radeon RX 6500 XT (radeonsi, navi24, LLVM 16.0.6, DRM 3.57, 6.7.4-arch1-1)/1] wgpu_test::occlusion_query::occlusion_query
       FLAKY 2/3 [   0.236s] wgpu-test::wgpu-test [Executed] [Gl/AMD Radeon RX 6500 XT (radeonsi, navi24, LLVM 16.0.6, DRM 3.57, 6.7.4-arch1-1)/1] wgpu_test::regression::issue_3457::pass_reset_vertex_buffer
       FLAKY 2/3 [   0.278s] wgpu-test::wgpu-test [Executed] [Gl/AMD Radeon RX 6500 XT (radeonsi, navi24, LLVM 16.0.6, DRM 3.57, 6.7.4-arch1-1)/1] wgpu_test::shader_primitive_index::draw_indexed
      TRY 3 SEGV [   0.727s] wgpu-test::wgpu-test [Executed] [Gl/AMD Radeon RX 6500 XT (radeonsi, navi24, LLVM 16.0.6, DRM 3.57, 6.7.4-arch1-1)/1] wgpu_test::mem_leaks::simple_draw_check_mem_leaks
      TRY 3 SEGV [   0.579s] wgpu-test::wgpu-test [Executed] [Gl/AMD Radeon RX 6500 XT (radeonsi, navi24, LLVM 16.0.6, DRM 3.57, 6.7.4-arch1-1)/1] wgpu_test::regression::issue_3349::multi_stage_data_binding
      TRY 3 SEGV [   0.552s] wgpu-test::wgpu-test [Executed] [Gl/AMD Radeon RX 6500 XT (radeonsi, navi24, LLVM 16.0.6, DRM 3.57, 6.7.4-arch1-1)/1] wgpu_test::scissor_tests::scissor_test_empty_rect
      TRY 3 SEGV [   0.504s] wgpu-test::wgpu-test [Executed] [Gl/AMD Radeon RX 6500 XT (radeonsi, navi24, LLVM 16.0.6, DRM 3.57, 6.7.4-arch1-1)/1] wgpu_test::scissor_tests::scissor_test_empty_rect_with_offset
      TRY 3 SEGV [   0.569s] wgpu-test::wgpu-test [Executed] [Gl/AMD Radeon RX 6500 XT (radeonsi, navi24, LLVM 16.0.6, DRM 3.57, 6.7.4-arch1-1)/1] wgpu_test::scissor_tests::scissor_test_full_rect
  ```
        
  </details>

- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

Edit:

**Additional changes**

* Removes perVertexStruct naming as requested below https://github.com/gfx-rs/wgpu/pull/5227#issuecomment-1936027091. For reference, this was originally added in https://github.com/gfx-rs/naga/pull/789.
 * Entry point specific code from `parse_function` was extracted and now runs after all functions are parsed. Since this includes the code that does gl_PerVertex builtin culling, we can be sure that any potential access has been registered.

Imberflur committed:
Remove unnecessary "gl_PerVertex" name check so unused builtins will be
handled even when this name is not in the input SPIRV.

Imberflur commented on naga/src/front/spv/function.rs:
```diff
@@ -376,34 +376,33 @@ impl<I: Iterator<Item = u32>> super::Frontend<I> {
+                            span,
+                        } = ty.inner
+                        {
+                            let mut new_members = None;
```
I refactored this a bit so that it only clones the members list if it is going to be modified

teoxoy commented:
This looks good, could we also take out the block below? I don't think we should be naming things if they don't have a name in the input.

https://github.com/gfx-rs/wgpu/blob/665c075fa01a79859e92b64cfc66b5c42ab20769/naga/src/front/spv/mod.rs#L5130-L5139

teoxoy commented:
> Would it be worth adding a test for this?

That would be great, thanks!

> If so, would that go under `naga/tests/in/spv`?

Yes.

> Also, I'm not very familiar with the overall structure of SPIRV nor with the spv frontend for naga. Can parse_function run more than once for a particular input?

SPIR-V does support multiple entry points per file but they are not common.

> If so, is it possible that later functions will be using the builtin but we remove the builtin label after not seeing any uses when processing the first function?

This is a good point, we might need to defer the removal.

Imberflur committed:
[spv-in] Remove providing "perVertexStruct" for certain unnamed inputs
since we probably shouldn't be naming things if they don't have a name
in the input.

As requested here: https://github.com/gfx-rs/wgpu/pull/5227#issuecomment-1936027091

Imberflur committed:
Add test case to naga/spv-in for an unnamed glPerVertex struct

Imberflur commented:
> This is a good point, we might need to defer the removal.

This seems to be the case, I was able to get spv-in to remove the builtin from position by moving positing setting out of the main function:
```glsl
#version 450

out gl_PerVertex {
    vec4 gl_Position;
    float gl_ClipDistance[1];
};

void builtin_usage() {
    gl_ClipDistance[0] = 1.0;
    gl_Position = vec4(
        (gl_VertexIndex == 0) ? -4.0 : 1.0,
        (gl_VertexIndex == 2) ? 4.0 : -1.0,
        0.0,
        1.0
    );
}

void main()
{
    builtin_usage();
}
```
```sh
glslc test.vert -o test.spv
cargo r --bin naga test.spv
```
```
Entry point main at Vertex is invalid:
	Vertex shaders must return a `@builtin(position)` output value
```

Imberflur commented:
> If so, is it possible that later functions will be using the builtin but we remove the builtin label after not seeing any uses when processing the first function?

I'm reading the code further and I realized that this section of code only runs for the entry point function, so the issue is not quite as bad as I make it seem here. It seems like it's only an issue if the entry point itself doesn't access the builtin while another function that the entry point calls does access it.

Imberflur committed:
Extract entry point processing from parse_function into an independent function.

Imberflur committed:
Defer entry-point processing so that all potential uses of builtin's are found before culling the unused ones

Imberflur committed:
Add test case for builtin only accessed in function parsed after the entry point function

Imberflur commented on naga/src/front/spv/function.rs:
```diff
@@ -292,278 +292,286 @@ impl<I: Iterator<Item = u32>> super::Frontend<I> {
-                named_expressions: crate::NamedExpressions::default(),
-                body: crate::Block::new(),
-            };
+            self.deferred_entry_points.push((ep, fun_id));
```
The diff here is a lot more readable with whitespace changes hidden:
![image](https://github.com/gfx-rs/wgpu/assets/2002109/1345ac14-96af-404c-b2da-898aacb6a738)

I extracted the entry point processing from this `if let` block into a separate `process_entry_point` function and then deferred that to be called after all functions are parsed.

Imberflur commented on naga/tests/snapshots.rs:
```diff
@@ -886,6 +886,12 @@ fn convert_spv_all() {
     );
+    convert_spv(
+        "unnamed-gl-per-vertex",
+        true,
```
Not sure if the `adjust_coordinate_space` should be `true`, but it might not matter either way for this test case?

teoxoy replied:
Yeah, I think any value should be fine.

Imberflur commented on naga/tests/in/spv/unnamed-gl-per-vertex.spvasm:
```diff
@@ -0,0 +1,72 @@
+;; Make sure that we don't have a validation error due to lacking capabilities
```
Trying to mimic the existing `.spvasm` files, I manually created this file with `spirv-dis unnamed-gl-per-vertex.spv > unnamed-gl-per-vertex.spvasm` and added some relevant comments in it. I don't know if there are any standard practices for adding naga tests here that I might have missed.

teoxoy replied:
The tests look great, thanks for adding the comments as well!

Imberflur reviewed (commented):
Notes for additional context

Imberflur commented:
> > Would it be worth adding a test for this?
>
> That would be great, thanks!

Added input spv to test for the original issue as well as another to test for the case of builtin's being accessed in a function parsed after the entry point.

>  This is a good point, we might need to defer the removal.

I ended up finding a way to defer this by postponing the entry point related processing to be after all function parsing. If this change is too much for this PR I can move it into a separate one.

> This looks good, could we also take out the block below? I don't think we should be naming things if they don't have a name in the input.

Removed this block.

teoxoy reviewed (approved):
Great stuff, thanks!

teoxoy commented:
One last thing, I would like to "rebase and merge" this PR instead of squashing it. Could you include the changes in the last commit ("[fix typo](https://github.com/gfx-rs/wgpu/pull/5227/commits/4e346622a04df3297df4e39e7693336d5fc52611)") in the appropriate commits that introduced those?

Imberflur commented:
> One last thing, I would like to "rebase and merge" this PR instead of squashing it. Could you include the changes in the last commit ("[fix typo](https://github.com/gfx-rs/wgpu/pull/5227/commits/4e346622a04df3297df4e39e7693336d5fc52611)") in the appropriate commits that introduced those?

done
